### PR TITLE
LUCENE-9524: Fix NPE in SpanWeight#explain when no scoring is require…

### PR DESCRIPTION
**Pull request by zacharymorn**
# Description
`SpanWeight#explain` uses `Similarity.SimScorer` to generate explanations, and may fail with NullPointerException when scoring is not needed and thus `Similarity.SimScorer` is set to null, such as matching for must not occur span query.

# Solution
The solution is to check for null `Similarity.SimScorer` in `SpanWeight#explain`, and provides alternative explanation.

# Tests
Added 1 integration test to `TestMemoryIndex` that was broken before the fix.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java [hunk 1]
```diff
@@ -618,4 +622,21 @@ public void testToStringDebug() {
         "fields=2, terms=2, positions=3", mi.toStringDebug());
   }
 
+  public void testExplain() throws ParseException, IOException {
```
Comment (jpountz) [inline lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java:625]: I don't think this is specific to MemoryIndex, so can you move this test to TestSpansExplanations?

Comment (zacharymorn) [reply lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java]: Done

File: lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java [hunk 2]
```diff
@@ -618,4 +622,21 @@ public void testToStringDebug() {
...
+    Query query = new ComplexPhraseQueryParser("content", analyzer).parse(queryString);
+
+    MemoryIndex memoryIndex = new MemoryIndex(true);
+    memoryIndex.addField("content", text, analyzer);
+
+    IndexSearcher searcher = memoryIndex.createSearcher();
+
+    TopDocs topDocs = searcher.search(query, 1);
+    ScoreDoc match = topDocs.scoreDocs[0];
+    searcher.explain(query, match.doc);
```
[the remark this answers is not in the document]
Comment (zacharymorn) [reply lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java]: I've tried this and it seems to be a bit more complicated to do. The problem here is that under the current test setup, the new explanation string code will only be exercised when there's a match between the (none-scoring) prohibited clause and the text.  So the matched topDoc that exercised that code path will actually not include that particular explanation as it must have been matching alternative clause instead.

Therefore, I'm considering two alternatives here. One would be to test `SpanWeight#explain` directly with null `SimScorer`; the other would be to setup another test to match span query without scoring, but I tried query with Occur.FILTER and MUST and see there's still a 1.0 boost associated and valid `SimScorer`. So I'm not sure if the latter approach is something actually achievable / meaningful to implement? 

Screen Shot 2020-10-14 at 7 26 18 PM

[the remark this answers is not in the document]
Comment (zacharymorn) [reply lucene/memory/src/test/org/apache/lucene/index/memory/TestMemoryIndex.java]: I've found a work-around in test and able to test with tighter condition, and assert for explanation message. Could you please let me know if this looks good to you?

File: lucene/core/build.gradle [hunk 3]
```diff
@@ -22,4 +22,5 @@ description = 'Lucene core library'
 dependencies {
   testImplementation project(':lucene:codecs')
   testImplementation project(':lucene:test-framework')
+  testImplementation project(':lucene:queryparser')
```
Comment (jpountz) [inline lucene/core/build.gradle:25]: we should be able to remove this now?

**Review by jpountz — APPROVED** [review]
Thanks, the new test looks good to me! I left a minor comment about a dependency that looks no longer needed.

File: lucene/core/build.gradle (hunk 3 above)
Comment (zacharymorn) [reply lucene/core/build.gradle]: Ops yes removed.

**Comment by jpountz** [comment]
Thanks @zacharymorn !

**Pull request merged**